feat: Add support for fetching cephfs mirroring status and lists to replication framework

UtkarshBhatthere opened the pull request:
```
root@primary:~# microceph replication list cephfs ist cephfs
+--------+-------------------------+-----------+
| VOLUME | RESOURCE                | PATH      |
+--------+-------------------------+-----------+
| vol    | /one                    | directory |
| vol    | /two                    | directory |
| vol    | /volumes/_nogroup/hpc   | subvolume |
| vol    | /volumes/openstack/ceph | subvolume |
+--------+-------------------------+-----------+
root@primary:~# microceph replication status cephfs vol
+--------------------------+
|          SUMMARY         |
+----------------+---------+
| Volume         | vol     |
| Resource Count | 4       |
| Peer Count     | 1       |
+----------------+---------+

+-------------+-------------------------+-------+--------------+---------------+---------------+
| REMOTE NAME | RESOURCE PATH           | STATE | SNAPS SYNCED | SNAPS DELETED | SNAPS RENAMED |
+-------------+-------------------------+-------+--------------+---------------+---------------+
| secondary   | /one                    | idle  |            3 |               0               |
| secondary   | /two                    | idle  |            1 |               0               |
| secondary   | /volumes/_nogroup/hpc   | idle  |            1 |               0               |
| secondary   | /volumes/openstack/ceph | idle  |            1 |               0               |
+-------------+-------------------------+-------+--------------+-------------------------------+
```

UtkarshBhatthere committed:
feat: Add support for fetching CephFS mirroring status

Signed-off-by: utkarsh bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
feat: Added Cephfs replication handler

Signed-off-by: utkarsh bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
temp: Added info logs

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
fix: Adds a patch for mirroring module from upstream

This patch is merged in squid and expected to land in the 19.2.4 point
release. Thus it is being added a temporary patch to microceph repo.

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
test: Modified functional tests to verify new commands

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

sabaini commented on microceph/ceph/cephfs_volume.go:
```diff
@@ -0,0 +1,155 @@
+	response.UngroupedSubVolumes = unsafe.Slice(
+		(*UngroupedSubvolume)(unsafe.SliceData(Subvolumes)),
+		len(Subvolumes),
+	)
```
Bit uneasy about the unsafety, could we do sth like the below instead?

```
response.UngroupedSubVolumes = make([]UngroupedSubvolume, len(Subvolumes))
for i, sv := range Subvolumes {
    response.UngroupedSubVolumes[i] = UngroupedSubvolume(sv)
}
```

Similarly below

sabaini replied:
Ack, gotcha. I view this as a tradeoff between performance / robustness -- with the unsafe construct we give up some static type checking which could impact future refactorings etc.

Is this an operation that you expect be run repeatedly / frequently? I'd advise to only go for the unsafe construct if we expect real world impact.

UtkarshBhatthere replied:
fair enough, I have modified it to follow the safer copy construct and prepare the slice.

sabaini commented on microceph/ceph/cephfs_volume.go:
```diff
@@ -0,0 +1,155 @@
+		args = []string{"fs", "subvolume", "getpath", volume, subvolume, subvolumegroup}
+	} else {
+		args = []string{"fs", "subvolume", "getpath", volume, subvolume}
+	}
```
Hm, is the logic here correct? If the subvolumegroup string is empty we use it to query?

sabaini commented on microceph/ceph/replication_cephfs.go:
```diff
@@ -0,0 +1,305 @@
+	return nil
+}
+
+// GetResourceState fetches the mirroring state for requested rbd pool/image.
```
Docstring copy paste error :-) 
Same with methods below

sabaini commented on microceph/cmd/microceph/replication_list_cephfs.go:
```diff
@@ -0,0 +1,84 @@
+
+	t := table.NewWriter()
+	t.SetOutputMirror(os.Stdout)
+	t.AppendHeader(table.Row{"Volume", "Resource", "Path"}, rowConfigAutoMerge)
```
Should these headers not be `Volume | Resource Path | Type` or so?

sabaini commented on microceph/cmd/microcephd/main.go:
```diff
@@ -122,7 +122,7 @@ func main() {
 	app.PersistentFlags().BoolVarP(&daemonCmd.global.flagHelp, "help", "h", false, "Print help")
 	app.PersistentFlags().BoolVar(&daemonCmd.global.flagVersion, "version", false, "Print version number")
-	app.PersistentFlags().BoolVarP(&daemonCmd.global.flagLogDebug, "debug", "d", false, "Show all debug messages")
+	app.PersistentFlags().BoolVarP(&daemonCmd.global.flagLogDebug, "debug", "d", true, "Show all debug messages")
```
Don't debug log in prod code

UtkarshBhatthere replied:
yeah my original changes were not rebased on your logger changes so i had to do this dance for debugging.

UtkarshBhatthere replied:
let me disable it again.

sabaini reviewed (changes requested):
Hey hey, first run through with some comments / requests -- thanks @UtkarshBhatthere 

Before we can merge we'd also need unit tests for this

UtkarshBhatthere commented on microceph/ceph/cephfs_mirror.go:
```diff
@@ -0,0 +1,239 @@
+	return "", nil
+}
+
+// Executes "help" on admin socket to verify the correctness.
```
ack

UtkarshBhatthere commented on microceph/ceph/cephfs_mirror.go:
```diff
@@ -0,0 +1,239 @@
+		}
+
+		full_sock_path := filepath.Join(run_path, socket.Name())
+		err = CheckFsMirrorHelperCommands(full_sock_path)
```
i will modify this to find the most recently created socket.

UtkarshBhatthere replied:
so apparently the most recent one is not a sure shot solution. We still end up iterating over a bunch of them.

sabaini replied:
Hm, did you add that TODO comment?

UtkarshBhatthere commented on microceph/ceph/cephfs_volume.go:
```diff
@@ -0,0 +1,155 @@
+	CephFsSubVolumeIndex              = 3 // "subvolume name"
+)
+
+func ListCephFSVolumes() ([]Volume, error) {
```
ack, added doc strings for all methods.

UtkarshBhatthere commented on microceph/ceph/cephfs_mirror.go:
```diff
@@ -0,0 +1,239 @@
+			if fs.Name == rh.Request.Volume {
+				volumeId = fs.FilesystemID
+				peers = make([]string, 0, len(fs.Peers))
+				logger.Errorf("filesystemID %d and peers %v", volumeId, fs.Peers)
```
ack.

UtkarshBhatthere committed:
fix: resolve review comments

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
fix: comment resolutions

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
fix: added timeout for list status

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
fix: Fix volume fetcher for missing ungrouped subvolumes

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

sabaini reviewed (approved):
Hey @UtkarshBhatthere lgtm in general, only thing missing is that TODO comment :-)

UtkarshBhatthere committed:
chore: Added a TODO for future

Added todo for future migration to ceph cli against the local admin sock
available from cephfs-mirror daemon.

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>
